cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -263,7 +264,12 @@ struct ValueConverter<TimestampType> {
           value = internal::PyDateTime_to_us(dt) - offset * 1000 * 1000;
           break;
         case TimeUnit::NANO:
-          value = internal::PyDateTime_to_ns(dt) - offset * 1000 * 1000 * 1000;
+          // Conversion to nanoseconds can overflow -> check multiply of microseconds
+          value = internal::PyDateTime_to_us(dt);
+          if (arrow::internal::MultiplyWithOverflow(value, 1000, &value)) {
+            return internal::InvalidValue(obj, "out of bounds for nanosecond resolution");
+          }
+          value -= offset * 1000 * 1000 * 1000;
```

Comment:
This subtraction could overflow as well, no?

Reply:
The offset is always a small number (time zone offsets are always less then 12 hours ?), so shouldn't give any problem?

Reply:
Ah, the offset itself won't overflow, but for the *subtraction* (as you actually commented ..) that could indeed happen in a small corner case: if your naive timezone is close to the lower or upper limit of the nanosecond range, applying the offset could indeed overflow.